components/layout_2020/fragment_tree/box_fragment.rs

```diff
@@ -238,30 +241,78 @@ impl BoxFragment {
+                overflow.max_x().max(scrollable_overflow.max_x()),
+                overflow.max_y().max(scrollable_overflow.max_y()),
+            );
+
+            let overflow_style = self.style.effective_overflow(self.base.flags);
+            if overflow_style.y == ComputedOverflow::Visible {
+                overflow.origin.y = overflow.origin.y.min(scrollable_overflow.origin.y);
+                overflow.size.height = bottom_right.y - overflow.origin.y;
+            }
+
+            if overflow_style.x == ComputedOverflow::Visible {
+                overflow.origin.x = overflow.origin.x.min(scrollable_overflow.origin.x);
+                overflow.size.width = bottom_right.x - overflow.origin.x;
+            }
+        }
+
+        // <https://drafts.csswg.org/css-overflow-3/#scrollable-overflow-region>
+        // > ...accounting for transforms by projecting each box onto the plane of
+        // > the element that establishes its 3D rendering context. [CSS3-TRANSFORMS]
+        // Both boxes and its scrollable overflow (if it is included) should be transformed accordingly.
+        //
+        // FIXME: We are supposed to handle perspective transform and 3d context, but it is yet to happen.
+        if self.style.has_transform_or_perspective(self.base.flags) {
+            if let Some(transform) =
+                self.calculate_transform_matrix(&self.border_rect().to_untyped())
+            {
+                if let Some(transformed_overflow_box) =
+                    transform.outer_transformed_rect(&overflow.to_webrender().to_rect())
+                {
+                    overflow =
+                        f32_rect_to_au_rect(transformed_overflow_box.to_untyped()).cast_unit();
+                }
+            }
         }
 
-        // https://www.w3.org/TR/css-overflow-3/#scrollable
-        // Only include the scrollable overflow of a child box if it has overflow: visible.
+        overflow
+    }
+
+    /// <https://drafts.csswg.org/css-overflow/#unreachable-scrollable-overflow-region>
+    /// > area beyond the scroll origin in either axis is considered the unreachable scrollable overflow region
+    ///
+    /// Return the clipped the scrollable overflow based on its scroll origin, determined by overflow direction.
+    /// By default the clipping rect is the padding box of the fragment, this will coincides with the scrollport
+    /// if the fragment is a scroll container.
+    pub fn reachable_scrolling_overflow_region(
+        &self,
+        clipping_rect: Option<PhysicalRect<Au>>,
+    ) -> PhysicalRect<Au> {
+        let scrolling_direction = self.style.overflow_direction();
         let scrollable_overflow = self.scrollable_overflow();
-        let bottom_right = PhysicalPoint::new(
-            overflow.max_x().max(scrollable_overflow.max_x()),
-            overflow.max_y().max(scrollable_overflow.max_y()),
-        );
+        let clipping_rect = clipping_rect.unwrap_or(self.padding_rect());
```

Comment:
Let's extend `clipping_rect` as:
- if scrolling direction is rightward, extend right edge of `clipping_rect` to infinity; otherwise extend the left edge to -infinity
- extend the top/bottom edge of `clipping_rect` similarly

Then the result is simply the intersection of `scrollable_overflow` and `clipping_rect`, and we can get rid of the `(_, _)` branch

And then we can extend the tests to all combinations of directions and writing modes :)

(You may consider parameterizing the tests: https://web-platform-tests.org/writing-tests/testharness.html#variants)

Reply:
We have a test for this [scrollWidthHeight-negative-margin-002.html](https://wpt.fyi/results/css/cssom-view/scrollWidthHeight-negative-margin-002.html?label=master&product=chrome%5Bexperimental%5D&product=edge%5Bexperimental%5D&product=firefox%5Bexperimental%5D&product=safari%5Bexperimental%5D&product=servo&aligned&q=scrollwidthheight-negative-margin-002.html%20) but the test name contains many duplicates because `writing-mode` is not included. I guess we could make a new test that specifically test transform and does not depend on these `cssText`.

Reply:
Well, there are many ways to parameterize tests, and any of them is fine.

Reply:
Added a test using the same logic as [scrollWidthHeight-negative-margin-002.html](https://wpt.fyi/results/css/cssom-view/scrollWidthHeight-negative-margin-002.html?label=master&product=chrome%5Bexperimental%5D&product=edge%5Bexperimental%5D&product=firefox%5Bexperimental%5D&product=safari%5Bexperimental%5D&product=servo&aligned&q=scrollwidthheight-negative-margin-002.html%20), but using transform (instead of margin), and limit the scope of the test.